src/meta/src/rpc/server.rs

```diff
@@ -171,315 +153,244 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
-        SourceManager::new(
-            barrier_scheduler.clone(),
-            catalog_manager.clone(),
-            fragment_manager.clone(),
+        start_leader_srv(
+            meta_store,
+            address_info,
+            max_heartbeat_interval,
+            opts,
+            leader_rx,
+            elect_coord,
+            svc_shutdown_rx,
         )
         .await
-        .unwrap(),
-    );
-
-    let barrier_manager = Arc::new(GlobalBarrierManager::new(
-        scheduled_barriers,
-        env.clone(),
-        cluster_manager.clone(),
-        catalog_manager.clone(),
-        fragment_manager.clone(),
-        hummock_manager.clone(),
-        source_manager.clone(),
-        meta_metrics.clone(),
-    ));
-
-    {
-        let source_manager = source_manager.clone();
-        tokio::spawn(async move {
-            source_manager.run().await.unwrap();
-        });
-    }
+        .expect("Unable to start leader services");
+    });
 
-    let stream_manager = Arc::new(
-        GlobalStreamManager::new(
-            env.clone(),
-            fragment_manager.clone(),
-            barrier_scheduler.clone(),
-            cluster_manager.clone(),
-            source_manager.clone(),
-            hummock_manager.clone(),
-        )
-        .unwrap(),
-    );
+    Ok((join_handle, svc_shutdown_tx))
+}
 
-    hummock_manager
-        .purge_stale(
-            &fragment_manager
-                .list_table_fragments()
+// TODO: repeat test 100 times. Start test with different sleep intervals and so on
+// Print the sleep intervals
+// use pseudo rand gen with seed
+mod tests {
+    use std::net::{IpAddr, Ipv4Addr};
+
+    use risingwave_common::util::addr::HostAddr;
+    use risingwave_pb::common::WorkerType;
+    use risingwave_rpc_client::MetaClient;
+    use tokio::time::sleep;
+
+    use super::*;
+
+    async fn setup_n_nodes(n: u16) -> Vec<(JoinHandle<()>, WatchSender<()>)> {
+        let meta_store = Arc::new(MemStore::default());
+
+        let mut node_controllers: Vec<(JoinHandle<()>, WatchSender<()>)> = vec![];
+        for i in 0..n {
+            // TODO: add pseudo random sleep here
+            let node = format!("node{}", i);
+            let err_msg = format!("Meta {} failed in setup", node);
+            let err_msg = err_msg.as_str();
+
+            let info = AddressInfo {
+                addr: node,
+                listen_addr: SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 1234 + i),
+                prometheus_addr: None,
+                dashboard_addr: None,
+                ui_path: None,
+            };
+            node_controllers.push(
+                rpc_serve_with_store(
+                    meta_store.clone(),
+                    info,
+                    Duration::from_secs(4), // What values do we want to use here
+                    2,                      // What values do we want to use here
+                    MetaOpts::test(true),   // True or false?
```

Comment:
```suggestion
                    MetaOpts::test(false),   // True or false?
```

Reply:
Tried that this afternoon... I seem to run into ` panicked at 'Leader sender dropped: RecvError(())', src/meta/src/rpc/server.rs:141:18`. Tried to await the handle of `rpc_serve_with_store`, but that one did not really work. Not sure what I am doing wrong.

Reply:
Found the issue: I was not passing on the handlers correctly, which is why the green thread dropped.